src/Kestrel.Core/Internal/Http/PathNormalizer.cs

```diff
@@ -14,32 +14,40 @@ public static class PathNormalizer
         private const byte ByteDot = (byte)'.';
 
-        public static string DecodePath(Span<byte> path, bool pathEncoded, string rawTarget, int queryLength)
+        public unsafe static string DecodePath(ReadOnlySpan<byte> path, bool pathEncoded, string rawTarget, int queryLength)
         {
             int pathLength;
             if (pathEncoded)
             {
                 // URI was encoded, unescape and then parse as UTF-8
                 // Disabling warning temporary
-                pathLength = UrlDecoder.DecodeInPlace(path);
-
-                // Removing dot segments must be done after unescaping. From RFC 3986:
-                //
-                // URI producing applications should percent-encode data octets that
-                // correspond to characters in the reserved set unless these characters
-                // are specifically allowed by the URI scheme to represent data in that
-                // component.  If a reserved character is found in a URI component and
-                // no delimiting role is known for that character, then it must be
-                // interpreted as representing the data octet corresponding to that
-                // character's encoding in US-ASCII.
-                //
-                // https://tools.ietf.org/html/rfc3986#section-2.2
-                pathLength = RemoveDotSegments(path.Slice(0, pathLength));
-
-                return GetUtf8String(path.Slice(0, pathLength));
+                fixed (byte* p = path)
+                {
+                    Span<byte> writablePath = new Span<byte>(p, path.Length);
+                    pathLength = UrlDecoder.DecodeInPlace(writablePath);
+
+                    // Removing dot segments must be done after unescaping. From RFC 3986:
+                    //
+                    // URI producing applications should percent-encode data octets that
+                    // correspond to characters in the reserved set unless these characters
+                    // are specifically allowed by the URI scheme to represent data in that
+                    // component.  If a reserved character is found in a URI component and
+                    // no delimiting role is known for that character, then it must be
+                    // interpreted as representing the data octet corresponding to that
+                    // character's encoding in US-ASCII.
+                    //
+                    // https://tools.ietf.org/html/rfc3986#section-2.2
+                    pathLength = RemoveDotSegments(writablePath.Slice(0, pathLength));
+
+                    return GetUtf8String(writablePath.Slice(0, pathLength));
+                }
             }
 
-            pathLength = RemoveDotSegments(path);
+            fixed (byte* p = path)
+            {
+                Span<byte> writablePath = new Span<byte>(p, path.Length);
+                pathLength = RemoveDotSegments(writablePath);
+            }
```

Comment:
These changes seem like a bug waiting to happen: accepting a `ReadOnlySpan` and then mutating it. 😦

Reply:
But the previous method accepted a `Span<byte>`, not a `ReadOnlySpan<byte>`.

Reply:
> These changes seem like a bug waiting to happen: accepting a ReadOnlySpan and then mutating it. 

Yeah, though to be clear, I'm not adding this, just "deobfuscating" it.

Reply:
Before it was taking ReadOnlySequence, unpacking it into a pointer (https://github.com/aspnet/KestrelHttpServer/blob/release/2.2/src/Kestrel.Core/Internal/Http/HttpParser.cs#L66) higher up the stack and creating the Span out of the pointer(https://github.com/aspnet/KestrelHttpServer/blob/release/2.2/src/Kestrel.Core/Internal/Http/HttpParser.cs#L134) before calling this method. At leas not all unsafeness in concentrated in one place.

Reply:
> But the previous method accepted a Span<byte>, not a ReadOnlySpan<byte>.

Upstream the data comes in as a ReadOnlySequence<byte> that was pinned into a Span. I've pushed that down to the one place where it was actually depending on Span.

Reply:
> higher up the stack

That actually seems better to me.  With this new variant, code all throughout the stack thinks it's working with something that's read-only when in reality something it calls indirectly deep down may mutate it.  At least with the previous approach, it was something up at the top it seems that did the conversion, and everything down the stack from there knew "this thing might be mutated by something I call".

Why does it start off as `ReadOnly` in the first place?  Are we missing a mutable variant, and is that causing us problems?

Also, I've not tracked it through the whole PR, and am wondering why it's changing as part of this.  Is it just ancillary "clean up", or did this need to happen as part of the rewrite with `BufferReader`?  If it's the latter, and BufferReader is handing out a `ReadOnlySpan` that then needs to be mutated, it makes me wonder if we have a functionality gap, e.g. if the input actually came in as a `Span<T>` and was then cast to a `ReadOnlySpan<T>` as part of handing it off to the `BufferReader`, and if code then relies on the parsing of `BufferReader` and wants the slices it gets back to be mutable, how would we support that?

Reply:
I think we should just potentially just stackalloc here, paths are super tiny and editing in place is a hack.

Reply:
> I think we should just potentially just stackalloc here, paths are super tiny and editing in place is a hack.

I'm going to do this right now to avoid this issue. I'll allocate via ArrayPool if strictly necessary (over 128?).

Reply:
Sure

Reply:
Ok, fixed. Strangely I'm getting even better performance numbers for 'LiveAspNet'.